File: src/data/businessCase.ts
import cloneDeep from 'lodash/cloneDeep';

import {
  BusinessCaseModel,
  ProposedBusinessCaseSolution
} from 'types/businessCase';
import {
  ApiLifecycleCostLine,
  ApiLifecyclePhase,
  CostData,
  LifecycleCosts,
  LifecyclePhaseKey,
  LifecycleSolution,
  LifecycleYears
} from 'types/estimatedLifecycle';

const yearsObject = {
  year1: '',
  year2: '',
  year3: '',
  year4: '',
  year5: ''
};

export const defaultEstimatedLifecycle: LifecycleCosts = {
  development: {
    label: 'Development',
    isPresent: true,
    type: 'primary',
    years: cloneDeep(yearsObject)
  },
  operationsMaintenance: {
    label: 'Operations and Maintenance',
    isPresent: true,
    type: 'primary',
    years: cloneDeep(yearsObject)
  },
  helpDesk: {
    label: 'Help desk/call center',
    isPresent: false,
    type: 'related',
    years: cloneDeep(yearsObject)
  },
  software: {
    label: 'Software licenses',
    isPresent: false,
    type: 'related',
    years: cloneDeep(yearsObject)
  },
  planning: {
    label: 'Planning, support, and professional services',
    isPresent: false,
    type: 'related',
    years: cloneDeep(yearsObject)
  },
  infrastructure: {
    label: 'Infrastructure',
    isPresent: false,
    type: 'related',
    years: cloneDeep(yearsObject)
  },
  oit: {
    label: 'OIT services, tools, and pilots',
    isPresent: false,
    type: 'related',
    years: cloneDeep(yearsObject)
  },
  other: {
    label: 'Other services, tools, and pilots',
    isPresent: false,
    type: 'related',
    years: cloneDeep(yearsObject)
  }
};

export const defaultProposedSolution = {
  title: '',
  summary: '',
  acquisitionApproach: '',
  targetContractAwardDate: '',
  targetCompletionDate: '',
  pros: '',
  cons: '',
  estimatedLifecycleCost: cloneDeep(defaultEstimatedLifecycle),
  costSavings: '',
  security: {
    isApproved: null,
    isBeingReviewed: '',
    zeroTrustAlignment: ''
  },
  hosting: {
    type: '',
    location: '',
    cloudStrategy: '',
    cloudServiceType: ''
  },
  hasUserInterface: '',
  workforceTrainingReqs: ''
};

export const businessCaseInitialData: BusinessCaseModel = {
  status: 'OPEN',
  systemIntakeId: '',
  requestName: '',
  projectAcronym: '',
  requester: {
    name: '',
    phoneNumber: ''
  },
  businessOwner: {
    name: ''
  },
  businessNeed: '',
  collaborationNeeded: '',
  currentSolutionSummary: '',
  cmsBenefit: '',
  priorityAlignment: '',
  successIndicators: '',
  responseToGRTFeedback: '',
  // proposedSolutions: [cloneDeep(defaultProposedSolution)],
  preferredSolution: cloneDeep(defaultProposedSolution),
  alternativeA: cloneDeep(defaultProposedSolution),
  alternativeB: cloneDeep(defaultProposedSolution),
  createdAt: '',
  updatedAt: ''
};

type lifecycleCostLinesType = Record<LifecycleSolution, LifecycleCosts>;

/**
 * This function tells us whether the parameter bizCaseSolution has been started
 * @param bizCaseSolution - an alternative solution case (e.g. A, B)
 */
export const solutionHasFilledFields = (
  bizCaseSolution: ProposedBusinessCaseSolution
) => {
  if (!bizCaseSolution) {
    return false;
  }

  const {
    title,
    summary,
    acquisitionApproach,
    targetContractAwardDate,
    targetCompletionDate,
    security,
    hosting,
    hasUserInterface,
    pros,
    cons,
    costSavings,
    workforceTrainingReqs,
    estimatedLifecycleCost
  } = bizCaseSolution;

  /** Whether requester has entered costs for required lifecycle cost categories */
  const hasRequiredPhaseCosts: boolean = !![
    ...Object.values(estimatedLifecycleCost.development.years),
    ...Object.values(estimatedLifecycleCost.operationsMaintenance.years)
  ].find(cost => !!cost);

  const hasRelatedCosts = !!Object.values(estimatedLifecycleCost).find(
    phase => phase.isPresent && phase.type === 'related'
  );

  return !!(
    title ||
    summary ||
    acquisitionApproach ||
    targetContractAwardDate ||
    targetCompletionDate ||
    security.isApproved ||
    security.isBeingReviewed ||
    security.zeroTrustAlignment ||
    hosting.type ||
    hosting.location ||
    hosting.cloudStrategy ||
    hosting.cloudServiceType ||
    hasUserInterface ||
    pros ||
    cons ||
    costSavings ||
    workforceTrainingReqs ||
    hasRequiredPhaseCosts ||
    hasRelatedCosts
  );
};

const LifecyclePhaseMap: Record<ApiLifecyclePhase, LifecyclePhaseKey> = {
  Development: 'development',
  'Operations and Maintenance': 'operationsMaintenance',
  'Help desk/call center': 'helpDesk',
  'Software licenses': 'software',
  'Planning, support, and professional services': 'planning',
  Infrastructure: 'infrastructure',
  'OIT services, tools, and pilots': 'oit',
  Other: 'other'
};

export const prepareBusinessCaseForApp = (
  businessCase: any
): BusinessCaseModel => {
  /** Lifecycle costs from API */
  const lifecycleCosts: ApiLifecycleCostLine[] =
    businessCase.lifecycleCostLines;

  /** Default lifecycle costs object for app */
  const lifecycleCostLines: lifecycleCostLinesType = {
    Preferred: cloneDeep(defaultEstimatedLifecycle),
    A: cloneDeep(defaultEstimatedLifecycle),
    B: cloneDeep(defaultEstimatedLifecycle)
  };

  /** Merge lifecycle costs from api with default lifecycle costs */
  lifecycleCosts.forEach(line => {
    const { phase, solution } = line;

    const phaseKey = LifecyclePhaseMap[phase];

    /** Cost object for current phase */
    const costObject: CostData = lifecycleCostLines[solution][phaseKey];

    // Mark cost phase as `isPresent`
    costObject.isPresent = true;

    /** Cost converted to string */
    const cost: string = line?.cost === null ? '' : line?.cost.toString();

    // Set cost for correct year within cost object
    costObject.years[`year${line.year}` as keyof LifecycleYears] = cost;
  });

  return {
    id: businessCase.id,
    euaUserId: businessCase.euaUserId,
    status: businessCase.status,
    systemIntakeId: businessCase.systemIntakeId,
    requestName: businessCase.projectName || '',
    projectAcronym: businessCase.projectAcronym || '',
    requester: {
      name: businessCase.requester || '',
      phoneNumber: businessCase.requesterPhoneNumber || ''
    },
    businessOwner: {
      name: businessCase.businessOwner || ''
    },
    businessNeed: businessCase.businessNeed || '',
    collaborationNeeded: businessCase.collaborationNeeded || '',
    currentSolutionSummary: businessCase.currentSolutionSummary || '',
    cmsBenefit: businessCase.cmsBenefit || '',
    priorityAlignment: businessCase.priorityAlignment || '',
    successIndicators: businessCase.successIndicators || '',
    responseToGRTFeedback: businessCase.responseToGRTFeedback || '',
    // proposedSolutions: [],
    preferredSolution: {
      title: businessCase.preferredTitle || '',
      summary: businessCase.preferredSummary || '',
      acquisitionApproach: businessCase.preferredAcquisitionApproach || '',
      targetContractAwardDate:
        businessCase.preferredTargetContractAwardDate || '',
      targetCompletionDate: businessCase.preferredTargetCompletionDate || '',
      pros: businessCase.preferredPros || '',
      cons: businessCase.preferredCons || '',
      costSavings: businessCase.preferredCostSavings || '',
      estimatedLifecycleCost: lifecycleCostLines.Preferred,
      security: {
        isApproved: businessCase.preferredSecurityIsApproved,
        isBeingReviewed: businessCase.preferredSecurityIsBeingReviewed || '',
        zeroTrustAlignment:
          businessCase.preferredSecurityZeroTrustAlignment || ''
      },
      hosting: {
        type: businessCase.preferredHostingType || '',
        location: businessCase.preferredHostingLocation || '',
        cloudStrategy: businessCase.preferredHostingCloudStrategy || '',
        cloudServiceType: businessCase.preferredHostingCloudServiceType || ''
      },
      hasUserInterface: businessCase.preferredHasUI || '',
      workforceTrainingReqs: businessCase.preferredWorkforceTrainingReqs || ''
    },
    alternativeA: {
      title: businessCase.alternativeATitle || '',
      summary: businessCase.alternativeASummary || '',
      acquisitionApproach: businessCase.alternativeAAcquisitionApproach || '',
      targetContractAwardDate:
        businessCase.alternativeATargetContractAwardDate || '',
      targetCompletionDate: businessCase.alternativeATargetCompletionDate || '',
      pros: businessCase.alternativeAPros || '',
      cons: businessCase.alternativeACons || '',
      costSavings: businessCase.alternativeACostSavings || '',
      estimatedLifecycleCost: lifecycleCostLines.A,
      security: {
        isApproved: businessCase.alternativeASecurityIsApproved,
        isBeingReviewed: businessCase.alternativeASecurityIsBeingReviewed,
        zeroTrustAlignment:
          businessCase.alternativeASecurityZeroTrustAlignment || ''
      },
      hosting: {
        type: businessCase.alternativeAHostingType || '',
        location: businessCase.alternativeAHostingLocation || '',
        cloudStrategy: businessCase.alternativeAHostingCloudStrategy || '',
        cloudServiceType: businessCase.alternativeAHostingCloudServiceType || ''
      },
      hasUserInterface: businessCase.alternativeAHasUI || '',
      workforceTrainingReqs:
        businessCase.alternativeAWorkforceTrainingReqs || ''
    },
    alternativeB: {
      title: businessCase.alternativeBTitle || '',
      summary: businessCase.alternativeBSummary || '',
      acquisitionApproach: businessCase.alternativeBAcquisitionApproach || '',
      targetContractAwardDate:
        businessCase.alternativeBTargetContractAwardDate || '',
      targetCompletionDate: businessCase.alternativeBTargetCompletionDate || '',
      pros: businessCase.alternativeBPros || '',
      cons: businessCase.alternativeBCons || '',
      costSavings: businessCase.alternativeBCostSavings || '',
      estimatedLifecycleCost: lifecycleCostLines.B,
      security: {
        isApproved: businessCase.alternativeBSecurityIsApproved,
        isBeingReviewed: businessCase.alternativeBSecurityIsBeingReviewed || '',
        zeroTrustAlignment:
          businessCase.alternativeBSecurityZeroTrustAlignment || ''
      },
      hosting: {
        type: businessCase.alternativeBHostingType || '',
        location: businessCase.alternativeBHostingLocation || '',
        cloudStrategy: businessCase.alternativeBHostingCloudStrategy || '',
        cloudServiceType: businessCase.alternativeBHostingCloudServiceType || ''
      },
      hasUserInterface: businessCase.alternativeBHasUI || '',
      workforceTrainingReqs:
        businessCase.alternativeBWorkforceTrainingReqs || ''
    },
    createdAt: businessCase.createdAt,
    updatedAt: businessCase.updatedAt
  };
};

/** Array of lifecycle costs formatted for API */
const formatLifecycleCostsForApi = (
  lifecycleCosts: LifecycleCosts,
  solution: LifecycleSolution
): ApiLifecycleCostLine[] => {
  /** Array of lifecycle cost objects */
  const lifecycleCostsArray: CostData[] = Object.values(lifecycleCosts);

  // Reformat lifecycle costs for API and return as array
  return lifecycleCostsArray.reduce((acc, costData) => {
    const { label, years, isPresent } = costData;

    // Skip phases without any costs entered
    if (!isPresent) return acc;

    /** Phase label formatted for API */
    const phase: ApiLifecyclePhase =
      label === 'Other services, tools, and pilots' ? 'Other' : label;

    /** Array of costs for current phase */
    const currentPhaseCosts: string[] = Object.values(years);

    /** Current phase costs reformatted for API */
    const formattedLifecycleCosts: ApiLifecycleCostLine[] =
      currentPhaseCosts.map((cost, index) => ({
        solution,
        phase,
        year: (index + 1).toString(),
        cost: cost ? parseFloat(cost) : null
      }));

    return [...acc, ...formattedLifecycleCosts];
  }, [] as ApiLifecycleCostLine[]);
};

export const prepareBusinessCaseForApi = (
  businessCase: BusinessCaseModel
): any => {
  const alternativeBExists = solutionHasFilledFields(
    // businessCase.alternativeB
    businessCase.alternativeB
  );

  const alternativeAExists = solutionHasFilledFields(
    // businessCase.alternativeA
    businessCase.alternativeA
  );

  /** Preferred, Alternative A, and Alternative B lifecycle costs formatted for API */
  const lifecycleCostLines: ApiLifecycleCostLine[] = [
    ...formatLifecycleCostsForApi(
      businessCase.preferredSolution.estimatedLifecycleCost,
      'Preferred'
    ),

    ...(alternativeAExists
      ? formatLifecycleCostsForApi(
          businessCase.alternativeA.estimatedLifecycleCost,
          'A'
        )
      : []),

    ...(alternativeBExists
      ? formatLifecycleCostsForApi(
          businessCase.alternativeB.estimatedLifecycleCost,
          'B'
        )
      : [])
  ];

  // Return Business Case and convert empty strings to null for backend validation
  return {
    ...(businessCase.id && {
      id: businessCase.id
    }),
    ...(businessCase.euaUserId && {
      euaUserId: businessCase.euaUserId
    }),
    status: businessCase.status,
    systemIntakeId: businessCase.systemIntakeId,
    projectName: businessCase.requestName || null,
    projectAcronym: businessCase.projectAcronym || null,
    requester: businessCase.requester.name || null,
    requesterPhoneNumber: businessCase.requester.phoneNumber || null,
    businessOwner: businessCase.businessOwner.name || null,
    businessNeed: businessCase.businessNeed || null,
    collaborationNeeded: businessCase.collaborationNeeded || null,
    currentSolutionSummary: businessCase.currentSolutionSummary || null,
    cmsBenefit: businessCase.cmsBenefit || null,
    priorityAlignment: businessCase.priorityAlignment || null,
    successIndicators: businessCase.successIndicators || null,
    responseToGRTFeedback: businessCase.responseToGRTFeedback || null,

    // Preferred solution
    preferredTitle: businessCase.preferredSolution.title || null,
    preferredSummary: businessCase.preferredSolution.summary || null,
    preferredAcquisitionApproach:
      businessCase.preferredSolution.acquisitionApproach || null,
    preferredTargetContractAwardDate:
      businessCase.preferredSolution.targetContractAwardDate || null,
    preferredTargetCompletionDate:
      businessCase.preferredSolution.targetCompletionDate || null,
    preferredSecurityIsApproved:
      businessCase.preferredSolution.security.isApproved,
    preferredSecurityisBeingReviewed:
      businessCase.preferredSolution.security.isBeingReviewed || null,
    preferredSecurityZeroTrustAlignment:
      businessCase.preferredSolution.security.zeroTrustAlignment || null,
    preferredHostingType: businessCase.preferredSolution.hosting.type || null,
    preferredHostingLocation:
      businessCase.preferredSolution.hosting.location || null,
    preferredHostingCloudStrategy:
      businessCase.preferredSolution.hosting.cloudStrategy || null,
    preferredHostingCloudServiceType:
      businessCase.preferredSolution.hosting.cloudServiceType || null,
    preferredHasUI: businessCase.preferredSolution.hasUserInterface || null,
    preferredPros: businessCase.preferredSolution.pros || null,
    preferredCons: businessCase.preferredSolution.cons || null,
    preferredCostSavings: businessCase.preferredSolution.costSavings || null,
    preferredWorkforceTrainingReqs:
      businessCase.preferredSolution.workforceTrainingReqs || null,

    // Alternative A
    alternativeATitle: businessCase.alternativeA.title || null,
    alternativeASummary: businessCase.alternativeA.summary || null,
    alternativeAAcquisitionApproach:
      businessCase.alternativeA.acquisitionApproach || null,
    alternativeATargetContractAwardDate:
      businessCase.alternativeA.targetContractAwardDate || null,
    alternativeATargetCompletionDate:
      businessCase.alternativeA.targetCompletionDate || null,
    alternativeASecurityIsApproved:
      businessCase.alternativeA.security.isApproved,
    alternativeASecurityisBeingReviewed:
      businessCase.alternativeA.security.isBeingReviewed || null,
    alternativeASecurityZeroTrustAlignment:
      businessCase.alternativeA.security.zeroTrustAlignment || null,
    alternativeAHostingType: businessCase.alternativeA.hosting.type || null,
    alternativeAHostingLocation:
      businessCase.alternativeA.hosting.location || null,
    alternativeAHostingCloudStrategy:
      businessCase.alternativeA.hosting.cloudStrategy || null,
    alternativeAHostingCloudServiceType:
      businessCase.alternativeA.hosting.cloudServiceType || null,
    alternativeAHasUI: businessCase.alternativeA.hasUserInterface || null,
    alternativeAPros: businessCase.alternativeA.pros || null,
    alternativeACons: businessCase.alternativeA.cons || null,
    alternativeACostSavings: businessCase.alternativeA.costSavings || null,
    alternativeAWorkforceTrainingReqs:
      businessCase.alternativeA.workforceTrainingReqs || null,

    // Alternative B
    alternativeBTitle: businessCase.alternativeB.title || null,
    alternativeBSummary: businessCase.alternativeB.summary || null,
    alternativeBAcquisitionApproach:
      businessCase.alternativeB.acquisitionApproach || null,
    alternativeBTargetContractAwardDate:
      businessCase.alternativeB.targetContractAwardDate || null,
    alternativeBTargetCompletionDate:
      businessCase.alternativeB.targetCompletionDate || null,
    alternativeBSecurityIsApproved:
      businessCase.alternativeB.security.isApproved || null,
    alternativeBSecurityisBeingReviewed:
      businessCase.alternativeB.security.isBeingReviewed || null,
    alternativeBSecurityZeroTrustAlignment:
      businessCase.alternativeB.security.zeroTrustAlignment || null,
    alternativeBHostingType: businessCase.alternativeB.hosting.type || null,
    alternativeBHostingLocation:
      businessCase.alternativeB.hosting.location || null,
    alternativeBHostingCloudStrategy:
      businessCase.alternativeB.hosting.cloudStrategy || null,
    alternativeBHostingCloudServiceType:
      businessCase.alternativeB.hosting.cloudServiceType || null,
    alternativeBHasUI: businessCase.alternativeB.hasUserInterface || null,
    alternativeBPros: businessCase.alternativeB.pros || null,
    alternativeBCons: businessCase.alternativeB.cons || null,
    alternativeBCostSavings: businessCase.alternativeB.costSavings || null,
    alternativeBWorkforceTrainingReqs:
      businessCase.alternativeB.workforceTrainingReqs || null,
    lifecycleCostLines
  };
};

export const hostingTypeMap: any = {
  cloud: 'Yes, in the cloud (AWS, Azure, etc.)',
  dataCenter: 'Yes, at a data center',
  none: 'No, hosting is not needed'
};

File: pkg/models/business_case.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
	"github.com/guregu/null"
)

// BusinessCaseStatus represents the status of a system intake
type BusinessCaseStatus string

// LifecycleCostPhase represents the phase of a lifecycle cost line
type LifecycleCostPhase string

// LifecycleCostSolution represents the solution associated with the line
type LifecycleCostSolution string

// LifecycleCostYear represents the year associated with the line
type LifecycleCostYear string

const (
	// BusinessCaseStatusOPEN captures enum value "OPEN"
	BusinessCaseStatusOPEN BusinessCaseStatus = "OPEN"
	// BusinessCaseStatusCLOSED captures enum value "CLOSED"
	BusinessCaseStatusCLOSED BusinessCaseStatus = "CLOSED"

	// LifecycleCostPhaseDEVELOPMENT captures enum value "Development"
	LifecycleCostPhaseDEVELOPMENT LifecycleCostPhase = "Development"
	// LifecycleCostPhaseOPERATIONMAINTENANCE captures enum value "Operations and Maintenance"
	LifecycleCostPhaseOPERATIONMAINTENANCE LifecycleCostPhase = "Operations and Maintenance"
	// LifecycleCostPhaseHELPDESK captures enum value "Help desk/call center"
	LifecycleCostPhaseHELPDESK LifecycleCostPhase = "Help desk/call center"
	// LifecycleCostPhaseSOFTWARE captures enum value "Software licenses"
	LifecycleCostPhaseSOFTWARE LifecycleCostPhase = "Software licenses"
	// LifecycleCostPhasePLANNING captures enum value "Planning, support, and professional services"
	LifecycleCostPhasePLANNING LifecycleCostPhase = "Planning, support, and professional services"
	// LifecycleCostPhaseINFRASTRUCTURE captures enum value "Infrastructure"
	LifecycleCostPhaseINFRASTRUCTURE LifecycleCostPhase = "Infrastructure"
	// LifecycleCostPhaseOIT captures enum value "OIT Services, tools, and pilots"
	LifecycleCostPhaseOIT LifecycleCostPhase = "OIT Services, tools, and pilots"
	// LifecycleCostPhaseOTHER captures enum value "Other"
	LifecycleCostPhaseOTHER LifecycleCostPhase = "Other"

	// LifecycleCostSolutionPREFERRED captures enum value "Preferred"
	LifecycleCostSolutionPREFERRED LifecycleCostSolution = "Preferred"
	// LifecycleCostSolutionA captures enum value "A"
	LifecycleCostSolutionA LifecycleCostSolution = "A"
	// LifecycleCostSolutionB captures enum value "B"
	LifecycleCostSolutionB LifecycleCostSolution = "B"

	// LifecycleCostYear1 captures enum value "1"
	LifecycleCostYear1 LifecycleCostYear = "1"
	// LifecycleCostYear2 captures enum value "2"
	LifecycleCostYear2 LifecycleCostYear = "2"
	// LifecycleCostYear3 captures enum value "3"
	LifecycleCostYear3 LifecycleCostYear = "3"
	// LifecycleCostYear4 captures enum value "4"
	LifecycleCostYear4 LifecycleCostYear = "4"
	// LifecycleCostYear5 captures enum value "5"
	LifecycleCostYear5 LifecycleCostYear = "5"
)

// EstimatedLifecycleCost is the model for the cost of an estimated lifecycle line in the Business Case.
type EstimatedLifecycleCost struct {
	ID             uuid.UUID             `json:"id"`
	BusinessCaseID uuid.UUID             `json:"business_case" db:"business_case"`
	Solution       LifecycleCostSolution `json:"solution"`
	Phase          *LifecycleCostPhase   `json:"phase"`
	Year           LifecycleCostYear     `json:"year"`
	Cost           *int64                `json:"cost"`
}

func (e EstimatedLifecycleCost) GetMappingKey() uuid.UUID {
	return e.BusinessCaseID
}

func (e EstimatedLifecycleCost) GetMappingVal() *EstimatedLifecycleCost {
	return &e
}

// EstimatedLifecycleCosts models a list of EstimatedLifecycleCost line items
type EstimatedLifecycleCosts []EstimatedLifecycleCost

// Scan implements the sql.Scanner interface
func (e *EstimatedLifecycleCosts) Scan(src interface{}) error {
	return json.Unmarshal(src.([]byte), e)
}

// BusinessCase is the model for the Business Case form.
type BusinessCase struct {
	ID                                  uuid.UUID          `json:"id"`
	EUAUserID                           string             `json:"euaUserId" db:"eua_user_id"`
	SystemIntakeID                      uuid.UUID          `json:"systemIntakeId" db:"system_intake"`
	Status                              BusinessCaseStatus `json:"status"`
	ProjectName                         null.String        `json:"projectName" db:"project_name"`
	ProjectAcronym                      null.String        `json:"projectAcronym" db:"project_acronym"`
	Requester                           null.String        `json:"requester"`
	RequesterPhoneNumber                null.String        `json:"requesterPhoneNumber" db:"requester_phone_number"`
	BusinessOwner                       null.String        `json:"businessOwner" db:"business_owner"`
	BusinessNeed                        null.String        `json:"businessNeed" db:"business_need"`
	CollaborationNeeded                 null.String        `json:"collaborationNeeded" db:"collaboration_needed"`
	CurrentSolutionSummary              null.String        `json:"currentSolutionSummary" db:"current_solution_summary"`
	CMSBenefit                          null.String        `json:"cmsBenefit" db:"cms_benefit"`
	PriorityAlignment                   null.String        `json:"priorityAlignment" db:"priority_alignment"`
	SuccessIndicators                   null.String        `json:"successIndicators" db:"success_indicators"`
	ResponseToGRTFeedback               null.String        `json:"responseToGRTFeedback" db:"response_to_grt_feedback"`
	PreferredTitle                      null.String        `json:"preferredTitle" db:"preferred_title"`
	PreferredSummary                    null.String        `json:"preferredSummary" db:"preferred_summary"`
	PreferredAcquisitionApproach        null.String        `json:"preferredAcquisitionApproach" db:"preferred_acquisition_approach"`
	PreferredTargetContractAwardDate    *time.Time         `json:"preferredTargetContractAwardDate" db:"preferred_target_contract_award_date"`
	PreferredTargetCompletionDate       *time.Time         `json:"preferredTargetCompletionDate" db:"preferred_target_completion_date"`
	PreferredSecurityIsApproved         null.Bool          `json:"preferredSecurityIsApproved" db:"preferred_security_is_approved"`
	PreferredSecurityIsBeingReviewed    null.String        `json:"preferredSecurityIsBeingReviewed" db:"preferred_security_is_being_reviewed"`
	PreferredZeroTrustAlignment         null.String        `json:"preferredZeroTrustAlignment" db:"preferred_zero_trust_alignment"`
	PreferredHostingType                null.String        `json:"preferredHostingType" db:"preferred_hosting_type"`
	PreferredHostingLocation            null.String        `json:"preferredHostingLocation" db:"preferred_hosting_location"`
	PreferredHostingCloudStrategy       null.String        `json:"preferredHostingCloudStrategy" db:"preferred_hosting_cloud_strategy"`
	PreferredHostingCloudServiceType    null.String        `json:"preferredHostingCloudServiceType" db:"preferred_hosting_cloud_service_type"`
	PreferredHasUI                      null.String        `json:"preferredHasUI" db:"preferred_has_ui"`
	PreferredPros                       null.String        `json:"preferredPros" db:"preferred_pros"`
	PreferredCons                       null.String        `json:"preferredCons" db:"preferred_cons"`
	PreferredCostSavings                null.String        `json:"preferredCostSavings" db:"preferred_cost_savings"`
	PreferredWorkforceTrainingReqs      null.String        `json:"preferredWorkforceTrainingReqs" db:"preferred_workforce_training_reqs"`
	AlternativeATitle                   null.String        `json:"alternativeATitle" db:"alternative_a_title"`
	AlternativeASummary                 null.String        `json:"alternativeASummary" db:"alternative_a_summary"`
	AlternativeAAcquisitionApproach     null.String        `json:"alternativeAAcquisitionApproach" db:"alternative_a_acquisition_approach"`
	AlternativeATargetContractAwardDate *time.Time         `json:"alternativeATargetContractAwardDate" db:"alternative_a_target_contract_award_date"`
	AlternativeATargetCompletionDate    *time.Time         `json:"alternativeATargetCompletionDate" db:"alternative_a_target_completion_date"`
	AlternativeASecurityIsApproved      null.Bool          `json:"alternativeASecurityIsApproved" db:"alternative_a_security_is_approved"`
	AlternativeASecurityIsBeingReviewed null.String        `json:"alternativeASecurityIsBeingReviewed" db:"alternative_a_security_is_being_reviewed"`
	AlternativeAZeroTrustAlignment      null.String        `json:"alternativeAZeroTrustAlignment" db:"alternative_a_zero_trust_alignment"`
	AlternativeAHostingType             null.String        `json:"alternativeAHostingType" db:"alternative_a_hosting_type"`
	AlternativeAHostingLocation         null.String        `json:"alternativeAHostingLocation" db:"alternative_a_hosting_location"`
	AlternativeAHostingCloudStrategy    null.String        `json:"alternativeAHostingCloudStrategy" db:"alternative_a_hosting_cloud_strategy"`
	AlternativeAHostingCloudServiceType null.String        `json:"alternativeAHostingCloudServiceType" db:"alternative_a_hosting_cloud_service_type"`
	AlternativeAHasUI                   null.String        `json:"alternativeAHasUI" db:"alternative_a_has_ui"`
	AlternativeAPros                    null.String        `json:"alternativeAPros" db:"alternative_a_pros"`
	AlternativeACons                    null.String        `json:"alternativeACons" db:"alternative_a_cons"`
	AlternativeACostSavings             null.String        `json:"alternativeACostSavings" db:"alternative_a_cost_savings"`
	AlternativeAWorkforceTrainingReqs   null.String        `json:"alternativeAWorkforceTrainingReqs" db:"alternative_a_workforce_training_reqs"`
	AlternativeBTitle                   null.String        `json:"alternativeBTitle" db:"alternative_b_title"`
	AlternativeBSummary                 null.String        `json:"alternativeBSummary" db:"alternative_b_summary"`
	AlternativeBAcquisitionApproach     null.String        `json:"alternativeBAcquisitionApproach" db:"alternative_b_acquisition_approach"`
	AlternativeBTargetContractAwardDate *time.Time         `json:"alternativeBTargetContractAwardDate" db:"alternative_b_target_contract_award_date"`
	AlternativeBTargetCompletionDate    *time.Time         `json:"alternativeBTargetCompletionDate" db:"alternative_b_target_completion_date"`
	AlternativeBSecurityIsApproved      null.Bool          `json:"alternativeBSecurityIsApproved" db:"alternative_b_security_is_approved"`
	AlternativeBSecurityIsBeingReviewed null.String        `json:"alternativeBSecurityIsBeingReviewed" db:"alternative_b_security_is_being_reviewed"`
	AlternativeBZeroTrustAlignment      null.String        `json:"alternativeBZeroTrustAlignment" db:"alternative_b_zero_trust_alignment"`
	AlternativeBHostingType             null.String        `json:"alternativeBHostingType" db:"alternative_b_hosting_type"`
	AlternativeBHostingLocation         null.String        `json:"alternativeBHostingLocation" db:"alternative_b_hosting_location"`
	AlternativeBHostingCloudStrategy    null.String        `json:"alternativeBHostingCloudStrategy" db:"alternative_b_hosting_cloud_strategy"`
	AlternativeBHostingCloudServiceType null.String        `json:"alternativeBHostingCloudServiceType" db:"alternative_b_hosting_cloud_service_type"`
	AlternativeBHasUI                   null.String        `json:"alternativeBHasUI" db:"alternative_b_has_ui"`
	AlternativeBPros                    null.String        `json:"alternativeBPros" db:"alternative_b_pros"`
	AlternativeBCons                    null.String        `json:"alternativeBCons" db:"alternative_b_cons"`
	AlternativeBCostSavings             null.String        `json:"alternativeBCostSavings" db:"alternative_b_cost_savings"`
	AlternativeBWorkforceTrainingReqs   null.String        `json:"alternativeBWorkforceTrainingReqs" db:"alternative_b_workforce_training_reqs"`
	CreatedAt                           *time.Time         `json:"createdAt" db:"created_at"`
	UpdatedAt                           *time.Time         `json:"updatedAt" db:"updated_at"`
	ArchivedAt                          *time.Time         `db:"archived_at"`

	// TODO: these fields are unused in GQL Schema but still exist as DB columns.
	// Write a migration to drop these columns
	// See https://jiraent.cms.gov/browse/EASI-1693
	InitialSubmittedAt *time.Time `json:"initialSubmittedAt" db:"initial_submitted_at"`
	LastSubmittedAt    *time.Time `json:"lastSubmittedAt" db:"last_submitted_at"`
}

// TODO: NJD - do I needto do something similar with ProposedSolutions?

// BusinessCaseWithCosts is a helper to allow for legacy REST code
// to function that combines lifecycle costs with the Business Case model
type BusinessCaseWithCosts struct {
	BusinessCase
	LifecycleCostLines EstimatedLifecycleCosts `json:"lifecycleCostLines" db:"lifecycle_cost_lines"`
}

// BusinessCases is the model for a list of Business Cases
type BusinessCases []BusinessCase

func (b BusinessCase) GetMappingKey() uuid.UUID {
	return b.SystemIntakeID
}
func (b BusinessCase) GetMappingVal() *BusinessCase {
	return &b
}
